File: Sorting/QuickSort.cpp
#include <iostream>
using namespace std;

int partition(int a[], int l, int h)
{
    int i, j, pivot, temp;
    pivot = a[h];
    i = l;
    for(j=l; j<h; j++)
    {
        if(a[j]<=pivot)
        {
            temp = a[j];
            a[j] = a[i];
            a[i] = temp;
            i++;
        }
    }
    temp = a[h];
    a[h] = a[i];
    a[i] = temp;
    return (i);
}

void quick_sort(int a[], int l, int h)
{
    int P;
    if(l < h)
    {
        P = partition(a, l, h);
        quick_sort(a, l, P-1);
        quick_sort(a, P+1, h);
    }
}

int main()
{
    int array[50];
    int i, n;
    cout<<"Enter number of elements in array: ";
    cin>>n;
    for(i=0; i<n; i++)
    {
        cout<<"Enter element at position "<<i<<" : ";
        cin>>array[i];
    }
    quick_sort(array, 0, n-1);
    cout<<"\nSorted Array: ";
    for(i=0; i<n; i++)
    {
        cout<<array[i]<<"  ";
    }
    
    return 0;
}

File: Algorithms/Sorting/Sorting_Java/QuickSort.java
import java.io.*;
import java.util.Scanner;

public class QuickSort{

	public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        System.out.println("Enter the size of the array:");
        int n = in.nextInt();

        // Input array
        int[] arr = new int[n];
        System.out.println("Enter the array elements (one by one): ");
        for (int i = 0; i < n; ++i) {
            arr[i] = in.nextInt();
        }

        // Apply QuickSort sort algo
        arr = quickSort(arr, 0, n - 1);

        // Print sorted array
        System.out.println("Sorted array elements: ");
        for (int i = 0; i < n; ++i) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static int[] quickSort(int[] arr, int l, int r) {
        if (l < r) {
            // partition the array w.r.t pivot element
            int p = partition(arr, l, r);

            // Divide the array
            quickSort(arr, l, p - 1);
            quickSort(arr, p + 1, r);
        }
        return arr;
    }

    public static int partition(int[] arr, int l, int r) {
        int i = l;
        // Choose last element as pivot element
        int pivot = arr[r];

        for (int j = l; j < r; ++j) {
            if (arr[j] <= pivot) {
                swap(arr, i, j);
                i++;
            }
        }
        swap(arr, i, r);
        return i;
    }


    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
